For this project, we will be cleaning and preparing a dataset of classified ads on the German eBay website.

The data set has been modified by:
- randomly selecting 50,000 entries to reduce computational effort
- dirtied to increase difficulty of project

Original data set and full documentation could be found [here](https://www.kaggle.com/orgesleka/used-cars-database/data).

In [1]:
## Reading data set ##
import pandas as pd
import numpy as np

f = open('autos.csv', encoding='Latin-1')
autos = pd.read_csv(f)

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
## Initial data set obeservations ##
print(autos.info())
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

There are 50,000 entries to the autos data set.

There are 20 columns of data for each entry, but 5 of which have null entries that we would need to look out for. The column with the lease non-null entries is 'notRepairedDamage', with 40,171 obejcts, missing almost a fifth of the dataset.

The data set is read as int64 and objects, there are columns such as 'price' that we could parse into other data type that is easier to work with.

Column names is styled in camelcase instead of snakecase.

## Cleaning Column Names

In [4]:
# editing specific column names for clarification

col_edits = {
    "yearOfRegistration": "registration_year",
    "monthOfRegistration": "registration_month",
    "notRepairedDamage": "unrepaired_damage",
    "dateCreated": "ad_created",
    "powerPS": "power_in_ps"
}

autos.rename(columns=col_edits, inplace=True)

# Using regular expression for conversion 
# from camel case to snake case for remaining columns
import re

autos.rename(
    columns=lambda col:re.sub(r'(?<!^)(?=[A-Z])', '_', col).lower(),
    inplace=True)

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_in_ps   model   odometer  registration_month fuel_type  \
0    manuell          158  andere  150,000km                   3       lpg   
1  automatik          286     7er  150,000km                   6    benzin   
2    manuell          102    golf   70,000km                   7    benzin   
3  automatik           71  fortwo   70,000km                   6    benzin   
4    manuell            0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We first edited column names of "yearOfRegistration", "monthOfRegistration", "notRepairedDamage", "dateCreated" and "powerPS" to something that more accurately describe the data stored in these columns, whilst making sure to submit edits in snake case.

For the remaining columns, we converted the labels from camel case to snake case because the latter is Python's preferred case choice.

## Initial Exploration and Cleaning

In [5]:
autos.describe(include='all') # implementing parameter to return both categorical and numeric columns

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox   power_in_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

The descriptive statistics of the data set revealed the following points that should be investigated further:
- Columns "seller" and "offer type" are both mostly made of one value, "privat" and "Angebot" respectively. These columns do not offer much information, so are candidates to be dropped.
- Column "nr_of_pictures" is consist to completely 0s, and documentation in original data set noted that it was a bug that happened when scrapping the data, hence is another candidate to be dropped.
- "registration_year" has minimum of 1000, which is probably a case of error input; the most frequent data is NaN, which is something also worth investigating.
- Several columns, including "registration_year", "registration_month" and "postal_code", should be reformatted as integer types when they are float types now; "price" and "odometer" should be reformatted as numeric dtype.

In [6]:
# Reformatting 'price' and 'odometer' as numeric dtype

autos["price"] = autos["price"].str.replace(r'[^0-9]', '').astype('int64')

autos["odometer"] = autos["odometer"].str.replace(r'[^0-9]', '').astype('int64')

# Relabelling "odometer"
autos.rename(columns={"odometer": "odometer_km"},
             inplace=True)

We found that there we non-numerical characters included in data for "price" and "odometer", we removed these characters and reformatted the data to numeric dtypes for computational convenience. We also renamed "odometer" to "odometer_km" so as to remind ourselves of the unit the data was entered as.

## Exploring the Odometer and Price Columns

In [7]:
autos["price"].unique().shape

(2357,)

In [8]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [9]:
autos["price"].value_counts().head().sort_index(ascending=True)

0       1421
500      781
1000     639
1500     734
2500     643
Name: price, dtype: int64

In [10]:
autos["odometer_km"].unique().shape

(13,)

In [11]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [12]:
autos["odometer_km"].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

We use several different techniques to explore the data in columns "price" and "odometer_km".

For price, we find that the min to be \$0 and max \$100,000,000, with the mean at \$9,840, the median is at \$2,950. From these statistics, it would appear that the higher end of the data is containing more outliers that should be removed, as the mean is much higher than the median and hence data is positively skewed. One method to reduce the impact of the data skew affecting our analysis to follow is to only keep the data within the IQR (inter quartile range), which is between \$1,100 and \$7,200. However, removing both the top and bottom 25% of data seems a bit exccessive, so I would propose to just remove the zero dollar cars because we are investigating car sales data and price zero provides poor analysis value. We find that there are 1,421 entries (2.8\% of total sample size) where the price is zero, confirming that removing these entries will not cause too much data to be lost. We will keep the upper boundary with the upper quartile.

For odometer, there are only 13 unique values for this varible, with the smallest value (10,000km) with 264 entries and no missing values, it would seem unnecessary to remove any after dropping more than 25% of the data with the price boundaries we've set.

In [13]:
# Removing price outliers
autos_new = autos[autos["price"].between(1, 7201)]

autos_new.describe()

price  registration_year   power_in_ps    odometer_km  \
count  36110.000000       36110.000000  36110.000000   36110.000000   
mean    2491.319302        2002.736804     98.424425  134583.633343   
std     1887.635564          65.466242    193.103145   32447.054519   
min        1.000000        1000.000000      0.000000    5000.000000   
25%      950.000000        1998.000000     60.000000  125000.000000   
50%     1990.000000        2001.000000     95.000000  150000.000000   
75%     3750.000000        2005.000000    125.000000  150000.000000   
max     7200.000000        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        36110.000000         36110.0  36110.000000  
mean             5.626475             0.0  49869.410053  
std              3.750233             0.0  25647.121453  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  29303.000000  
50%              5.000000             0.0  48455.000000  
75%              9.000000             0.0  69126.000000  
max             12.000000             0.0  99998.000000

After removing the outliers, the new mean price is at \$2,491 and median is at \$1,990, which is closer together, suggesting a much more evenly distributed data set.

Odometer still kept its min and max values, however, the mean increased slightly.

The new data set contains 36,110 entries 72% of the original data.

## Exploring the date columns

In [14]:
# Exploring "date_crawled", "ad_created" and "last_seen"
autos_new[["date_crawled", "ad_created", "last_seen"]].describe()

date_crawled           ad_created            last_seen
count                 36110                36110                36110
unique                35188                   64                31023
top     2016-03-25 19:57:10  2016-04-03 00:00:00  2016-04-07 06:17:27
freq                      3                 1378                    6

Since the date columns are stored as objects, the methods we used to explore numeric dtypes are not really appropriate. Hence, we will adjust our method parameters to provide a better representation of the data.

However, from the normal describe method, we can see that both "date_crawled" and "last_seen" have large numbers of unique values, hence analysing the data via frequency table is not going to be very productive. As such, we will find the dates which these two columns range, and only conduct the frequency analysis on the "ad_created" column.

In [15]:
print('"date_crawled" range:',
      autos_new["date_crawled"].min(),
      "to",
      autos_new["date_crawled"].max())
print('"last_seen" range:',
      autos_new["last_seen"].min(),
      "to",
      autos_new["last_seen"].max())

"date_crawled" range: 2016-03-05 14:06:30 to 2016-04-07 14:36:44
"last_seen" range: 2016-03-05 14:45:46 to 2016-04-07 14:58:50


In [16]:
autos_new["ad_created"].value_counts(normalize=True, dropna=False, ascending=False).head().sort_index()

2016-03-12 00:00:00    0.037081
2016-03-20 00:00:00    0.037524
2016-03-21 00:00:00    0.037635
2016-04-03 00:00:00    0.038161
2016-04-04 00:00:00    0.037303
Name: ad_created, dtype: float64

It would seem that the data is between 5th March, to 7th April, 2016.

It is more difficult to see a pattern in the dates of "ad_created", though we could probably infer that Sundays and Mondays are popular dates of the week for upload.

In [17]:
autos_new["registration_year"].describe()

count    36110.000000
mean      2002.736804
std         65.466242
min       1000.000000
25%       1998.000000
50%       2001.000000
75%       2005.000000
max       9999.000000
Name: registration_year, dtype: float64

There are some obvious odd values in the data as the min registration year is 1000, whilst the max is 9999.

# Dealing with Incorrect Registration Year Data

In [18]:
# Finding number of rows that falls out of the range 1900 and 2016
len(autos_new["registration_year"]) - len(autos_new[autos_new["registration_year"].between(1900,2016)])

1631

Since we know that all listings happened in 2016, hence any cars with registration years after such is definitely inaccurate. However, determining the earliest valid year is an exercise more open to discretion.

We also know that the number of rows that falls out of the 1900-2016 interval is 1631, which is 4.5\% of the adjusted data set. This is not too much of a data lost.

Hence, we will set the upper bound as 2016 and lower bound as 1900.

In [19]:
# Dropping data out side of interval
autos_new = autos_new[autos_new["registration_year"].between(1900,2016)]

autos_new["registration_year"].value_counts(normalize=True)

2000    0.089127
1999    0.082630
2001    0.073494
2003    0.071522
2002    0.068215
2005    0.067519
1998    0.066968
2004    0.066794
1997    0.055686
2006    0.052438
1996    0.039067
2007    0.037327
1995    0.034920
2016    0.031584
2008    0.028423
2009    0.024914
1994    0.017576
1993    0.011630
2010    0.011456
1992    0.009658
1990    0.009368
1991    0.009049
2011    0.006961
2012    0.004989
1989    0.004524
1988    0.003364
1985    0.002030
1980    0.001972
1987    0.001827
2013    0.001798
          ...   
1978    0.000667
1970    0.000609
1981    0.000551
1977    0.000377
1960    0.000377
1972    0.000377
1979    0.000377
1973    0.000319
1971    0.000261
1969    0.000232
1966    0.000203
1976    0.000203
1964    0.000174
1965    0.000174
1974    0.000174
1975    0.000174
1967    0.000174
1910    0.000145
1968    0.000145
1962    0.000087
1958    0.000087
1963    0.000058
1956    0.000058
1934    0.000058
1937    0.000058
1950    0.000058
1961    0.000029
1953    0.0000

There are now 66 unique registration years and there are the greates number of cars registered in the 1990s and 2000s for this data set.

## Exploring Price by Brand

In [20]:
# Identifying brands to aggregate over
autos_new["brand"].unique().shape

(40,)

In [21]:
autos_new["brand"].value_counts(normalize=True).head(20)

volkswagen       0.211665
opel             0.131529
bmw              0.086110
ford             0.081238
mercedes_benz    0.080426
audi             0.063836
renault          0.059601
peugeot          0.036892
fiat             0.031468
seat             0.019925
smart            0.017779
mazda            0.017141
citroen          0.016039
nissan           0.015836
skoda            0.014966
toyota           0.013806
hyundai          0.010093
mitsubishi       0.009919
volvo            0.009600
honda            0.008730
Name: brand, dtype: float64

In [22]:
# Selecting brands fulfilling condition
freq_tab = autos_new["brand"].value_counts(normalize=True)

fulfill_num = sum(freq_tab.between(0.05, 1))

brands = freq_tab.head(fulfill_num).index.values.tolist()

We find that there are 40 brands, however, there are still obvious more popular brands as there are only a handful of brands that have over 5\% of data.

We will use the threshold of 5\% and above of total values to aggregate our data with and created a list to store the brands that fulfills this criteria.

In [28]:
# Finding mean prices of selected brands
brand_price = dict()

for b in brands:
    brand_price[b] = autos_new.loc[autos_new["brand"] == b, "price"].mean()

brand_price

{'audi': 2938.815992730577,
 'bmw': 3074.8002694509937,
 'ford': 2026.3227418779009,
 'mercedes_benz': 3146.6065632888567,
 'opel': 2029.8937155457552,
 'renault': 1795.160097323601,
 'volkswagen': 2462.110715264456}

We find that there are 7 brands of cars that have data size that exceeds 5\% of total data.

From the aggregate price of these 7 brands, we find that there is a distinct price gap occurring:
- Mercedes Benz, BMW and Audi are most expensive, with the highest average price at \$3,146.61, \$3,074.80 and \$2,938.82 respectively.
- Ford, Opel and Renault are less expensive
- Volkswagen is in between

## Storing Aggregate Data in a DataFrame

In [30]:
# Finding mean mileage of selected brands
brand_mile = dict()

for b in brands:
    brand_mile[b] = autos_new.loc[autos_new["brand"] == b, "odometer_km"].mean()

brand_mile

{'audi': 145433.89368468878,
 'bmw': 143482.65409228698,
 'ford': 131212.0671188861,
 'mercedes_benz': 141871.61918499818,
 'opel': 134997.79492833518,
 'renault': 132284.67153284673,
 'volkswagen': 138959.98903809264}

In [42]:
# Creating dataframe to store mean values
bp = pd.Series(brand_price)
bm = pd.Series(brand_mile)

# df = pd.DataFrame(data = {'mean_price': bp, 'mean_mileage': bm})
df = pd.DataFrame(bp, columns=['mean_price'])
df['mean_mileage'] = bm
df

mean_price   mean_mileage
audi           2938.815993  145433.893685
bmw            3074.800269  143482.654092
ford           2026.322742  131212.067119
mercedes_benz  3146.606563  141871.619185
opel           2029.893716  134997.794928
renault        1795.160097  132284.671533
volkswagen     2462.110715  138959.989038

The more expensive brands actually have higher average mileages than other brands.